cmd/joy/release.go

```diff
@@ -258,3 +266,111 @@ func NewReleaseRenderCmd() *cobra.Command {
 	cmd.Flags().BoolVar(&color, "color", term.IsTerminal(int(os.Stdout.Fd())), "toggle output with color")
 	return cmd
 }
+
+func NewGitQueryCommand(command string) *cobra.Command {
+	var (
+		source string
+		target string
+	)
+	cmd := &cobra.Command{
+		Use:  "git-" + command,
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cfg := config.FromContext(cmd.Context())
+
+			cat, err := catalog.Load(catalog.LoadOpts{Dir: cfg.CatalogDir})
+			if err != nil {
+				return err
+			}
+
+			release := args[0]
+
+			var sourceRelease, targetRelease *v1alpha1.Release
+
+			for _, cross := range cat.Releases.Items {
+				if cross.Name != release {
+					continue
+				}
+				for _, rel := range cross.Releases {
+					if rel == nil {
+						continue
+					}
+					switch rel.Environment.Name {
+					case source:
+						sourceRelease = rel
+					case target:
+						targetRelease = rel
+					}
+				}
+				break
+			}
+
+			if sourceRelease == nil {
+				return fmt.Errorf("no source found")
+			}
+			if targetRelease == nil {
+				return fmt.Errorf("no target found")
+			}
+
+			repoCache := filepath.Join(cfg.JoyCache, "src")
+			if err := os.MkdirAll(repoCache, 0o755); err != nil {
+				return fmt.Errorf("failed to ensure repo cache: %w", err)
+			}
+
+			var (
+				project = sourceRelease.Project
+				repo    = sourceRelease.Project.Spec.Repository
+			)
+
+			if _, err := os.Stat(filepath.Join(repoCache, project.Name)); err != nil {
+				if !errors.Is(err, os.ErrNotExist) {
+					return err
+				}
+
+				clone := exec.Command("gh", "repo", "clone", repo, project.Name)
+				clone.Stdout = os.Stdout
+				clone.Stderr = os.Stderr
+				clone.Dir = repoCache
+
+				if err := clone.Run(); err != nil {
+					return fmt.Errorf("failed to clone project: %w", err)
+				}
+			}
+
+			repoDir := filepath.Join(repoCache, project.Name)
+
+			getRevision := func(version string) string {
+				version = "v" + version
+				if branch := semver.Prerelease(version); branch != "" {
+					return branch
+				}
+				return path.Join(project.Spec.RepositorySubdirectory, version)
+			}
+
+			pull := exec.Command("git", "pull")
```

Comment:
Instead of a pull, I would do a `git fetch --tags`. That way the state of working tree is inconsequential.

Reply:
Also the `--tags` is important to fetch new tags.

Reply:
Probably need to do both a `git fetch` plus `git fetch --tags` since we need to be able to pull branches as well for environments that have PRs deployed to them

Reply:
Only `git fetch --tags` is required. It fetches more than just tags. That being said, we only care about tags.

Reply:
👍